rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,304 @@
+  /** A more detailed description of the service status. */
+  detail?: string;
+  /** A URL to open in a new tab about how to resolve or troubleshoot the problem */
+  documentationUrl?: string;
+  /** Any JSON-serializable data to be included in the HTTP API response. */
+  meta?: object;
+}
+
+/**
+ * The current "level" of availability of a service.
+ */
+enum ServiceStatusLevel {
+  available,   // everything is working!
+  degraded,    // some features may not be working
+  unavailable, // the service is unavailable, but other functions should work
+  fatal        // block all user functions and display the status page
+}
+
+/**
+ * Status of core services.
+ */
+interface CoreStatus {
+  elasticsearch: ServiceStatus;
+  http: ServiceStatus;
+  savedObjects: ServiceStatus;
+  uiSettings: ServiceStatus;
+  metrics: ServiceStatus;
+}
+```
+
+## API Design
+
+### Plugin API
+
+```ts
+/**
+ * The API exposed to plugins on CoreSetup.status
+ */
+interface StatusSetup {
+  /**
+   * Allows a plugin to specify a custom status dependent on its own criteria.
+   * Completely overrides the default inherited status.
+   */
+  set(status$: Observable<ServiceStatus>): Observable<ServiceStatus>;
+
+  /**
+   * Current status for all Core services.
+   */
+  core$: Observable<CoreStatus>;
+
+  /**
+   * Current status for all dependencies of the current plugin.
+   * Each key of the `Record` is a plugin id.
+   */
+  plugins$: Observable<Record<string, ServiceStatus>>;
+
+  /**
+   * The default status of the plugin as specified in the "Status inheritance"
+   * section below.
+   * 
+   * Useful when overriding the defaults with `set`.
+   */
+  inherited$: Observable<ServiceStatus>;
+}
+```
+
+### HTTP API
+
+The HTTP endpoint should return basic information about the Kibana node as well as the overall system status and the status of each individual system.
+
+This API does not need to include UI-specific details like the existing API such as `uiColor` and `icon`.
+
+```ts
+/**
+ * Response type for the endpoint: GET /api/status
+ */
+interface StatusResponse {
+  /** server.name */
+  name: string;
+  /** server.uuid */
+  uuid: string;
+  /** Currently exposed by existing status API */
+  version: {
+    number: string;
+    build_hash: string;
+    build_number: number;
+    build_snapshot: boolean;
+  };
+  /** Similar format to existing API, but slightly different shape */
+  status: {
+    overall: ServiceStatus;
+    core: Record<string, ServiceStatus>;
+    plugins: Record<string, ServiceStatus>;
+  }
+}
+```
+
+## Behaviors
+
+### Levels
+
+Each member of the `ServiceStatusLevel` enum has specific behaviors associated with it:
+- **`available`**:
+  - All endpoints and apps associated with the service are accessible
+- **`degraded`**:
+  - All endpoints and apps are available by default
+  - Some APIs may return `503 Unavailable` responses. This must be implemented directly by the service.
```

Comment:
What about plugin API exposed via contract?

Reply:
I hesitate from making any error throwing automatic at the JS API level, though, yes, I could specify here that calling a contract API for a >= degraded plugin may result in an exception.

My thinking here is that if your plugin is inheriting its status from its dependencies, then we are already throwing errors on the HTTP level so that these underlying contract APIs shouldn't be called at all.

Of course, this doesn't cover all cases. Some contract APIs may be called outside of the scope of an HTTP request. I just don't want too much magic, at least on this first pass. If we deem it useful or necessary later, I'm open to amending this RFC. WDYT?

Reply:
Yeah, a minimum level of magic is a must. Could you add a comment that the platform doesn't prevent plugins from calling degraded API via plugin contract?